src/xcoobee/sdk/Bees.js

```diff
@@ -1,247 +1,254 @@
+   * @property {boolean} result.page_info.has_next_page - Flag indicating whether there is
+   *   another page of data to may be fetched.
+   * @property {string} result.page_info.end_cursor - The end cursor.
+   *
+   * @throws {XcooBeeError}
+   */
+  async listBees(searchText = null, config = null) {
+    this._assertValidState();
+
+    const fetchPage = async (apiCfg, params) => {
+      const { apiKey, apiSecret, apiUrlRoot } = apiCfg;
+      const { after, limit, searchText: search } = params;
+      const apiAccessToken = await this._.apiAccessTokenCache.get(apiUrlRoot, apiKey, apiSecret);
+      const beesPage = await BeesApi.bees(apiUrlRoot, apiAccessToken, search, after, limit);
+      return beesPage;
+    };
+    const apiCfg = SdkUtils.resolveApiCfg(config, this._.config);
+    const params = { searchText };
+
+    return SdkUtils.startPaging(fetchPage, apiCfg, params);
+  }
+
+  /**
+   *
+   * @async
+   * @param {Array<Object<string, Object>>} bees - A mapping of bee names to bee parameters.
+   * @param {string} bees[].key - The bee name.  A 'transfer' bee will be ignored.
+   * @param {Object} bees[].value - The bee parameters.
+   * @param {Object} options - The bee take off options.
+   * @param {Object} options.process -
+   * @param {Array<{value: any, name: string}>} [options.custom] - custom properties
+   * @param {Array<email|XcooBeeId>} [options.process.destinations] -
+   * @param {string[]} options.process.fileNames -
+   * @param {string} [options.process.userReference] -
+   * @param {TransactionSubscriptions} [subscriptions]
+   * @param {Config} [config] - If specified, the configuration to use instead of the
+   *   default.
+   *
+   * @returns {Promise<SuccessResponse | ErrorResponse>} - The response.
+   * @property {number} code - The response status code.
+   * @property {Error} [error] - The response error if status is not successful.
+   * @property {string} [error.message] - The error message.
+   * @property {string} request_id - The ID of the request generated by the XcooBee
+   *   system.
+   * @property {Object} [result] - The result of the response if status is successful.
+   * @property {string} [result.ref_id] - A reference ID generated by the XcooBee
+   *   system.
+   *
+   * @throws {XcooBeeError}
+   */
+  async takeOff(bees, options, subscriptions, config = null) {
+    this._assertValidState();
+    const apiCfg = SdkUtils.resolveApiCfg(config, this._.config);
+    const { apiKey, apiSecret, apiUrlRoot } = apiCfg;
+
+    const directiveInput = {
+      filenames: options.process.fileNames,
+      user_reference: options.process.userReference || null,
+    };
+
+    if (subscriptions) {
+      // TODO: validate subscriptions
+      directiveInput.subscriptions = subscriptions;
+    }
+
+    if (
+      options.custom
+      && Array.isArray(options.custom)
+      && options.custom.length
+    ) {
+      directiveInput.custom = options.custom.slice();
```

Comment:
What's the reason behind making a copy of the custom props?

Reply:
it is a good practice to consider arguments immutable, I`m making a copy so that it won`t be changed by accident from inside of the function

Reply:
I agree that it is good practice to treat arguments as immutable. However, `slice` will only make a shallow copy. The items (the object containing the name value pairs) in the array are still mutable.

Fortunately, we pass the `directiveInput` to code we are in control of. So, I am not too worried about mutating arguments by accident. As the authors of the SDK and the underlying API code, I wouldn't bother. If we were passing the arguments to a third-party library, then it might be more important if it was an argument that we needed to re-use after the function call. If the user of our SDK is worried we will mutate their arguments, **they** could be the ones making the copies and not us.

So, you could either
1) Not worry about it.
2) Not worry about it, but put in checks in the unit/integration tests to ensure arguments have not been mutated (could pass in a frozen argument (`Object.freeze`) in strict mode. If there is a mutation, then an error would occur.)
3) Use defensive programming by making copies of mutable objects. Just be sure to make copies for everything that is mutable.

I'll approve and support your decision with whichever strategy you want to take. But, if you want to do `3` with the defensive programming, then I think we should handle it throughout the entire project, not just one function here and there. Please raise an issue if you want to do `3`. I am not sure exactly how Bilal wants that done. I usually talk to him first. Then either create a Trello Card or add a GitHub issue to the repo.

Reply:
I would tend to narrow down to 1 or 2

Reply:
We can also make sure that ESLint is configured to error when we do mutate an argument too.